server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java

```diff
@@ -347,66 +346,145 @@ protected boolean lessThan(IteratorAndCurrent a, IteratorAndCurrent b) {
             }
         }
 
-        mergeClosestBuckets(reducedBuckets, numClusters, reduceContext); // Reduce to <code>numClusters</code> buckets
-
-        orderBucketsByXContentKey(reducedBuckets, reduceContext);
+        mergeBucketsIfNeeded(reducedBuckets, numClusters, reduceContext);
 
         return reducedBuckets;
     }
 
+
+    class BucketRange{
+        int startIdx;
+        int endIdx;
+
+        /**
+         * These are optional utility fields
+         * They're useful for determining whether buckets should be merged
+         */
+        double min;
+        double max;
+        double centroid;
+        long docCount;
+
+        public void mergeWith(BucketRange other){
+            startIdx = Math.min(startIdx, other.startIdx);
+            endIdx = Math.max(endIdx, other.endIdx);
+
+            if(docCount + other.docCount > 0) {
+                // Avoids div by 0 error. This condition could be false if the optional docCount field was not set
+                centroid = ((centroid * docCount) + (other.centroid * other.docCount)) / (docCount + other.docCount);
+                docCount += other.docCount;
+            }
+            min = Math.min(min, other.min);
+            max = Math.max(max, other.max);
+        }
+    }
+
     /**
-     * Merges the two closest buckets together. Distance is determined by centroid comparison.
-     * Requires: <code>buckets</code> is sorted in ascending order by centroid.
+     * For each range {startIdx, endIdx} in <code>ranges</code>, all the buckets in that index range
+     * from <code>buckets</code> are merged, and this merged bucket replaces the entire range.
      */
-    void mergeClosestBuckets(List<Bucket> buckets, int targetNumBuckets, ReduceContext reduceContext){
-        while (buckets.size() > targetNumBuckets) {
-            // Find two closest buckets
-            int closest_idx = 0; // After this loop, (closest_idx, closest_idx + 1) will be the 2 closest buckets
-            double smallest_distance = Double.POSITIVE_INFINITY;
+    private void mergeBucketsWithPlan(List<Bucket> buckets, List<BucketRange> plan, ReduceContext reduceContext){
+        for(int i = plan.size() - 1; i >= 0; i--) {
+            BucketRange range = plan.get(i);
+            int endIdx = range.endIdx;
+            int startIdx = range.startIdx;
+
+            if(startIdx == endIdx) continue;
+
+            List<Bucket> to_merge = new ArrayList<>();
+            for(int idx = endIdx; idx > startIdx; idx--){
+                to_merge.add(buckets.get(idx));
+                buckets.remove(idx);
+            }
+            to_merge.add(buckets.get(startIdx)); // Don't remove the startIdx bucket because it will be replaced by the merged bucket
+
+            reduceContext.consumeBucketsAndMaybeBreak(1); // NOCOMMIT: Is this correct?
```

Comment:
I think it'd be more correct to `reduceContext.consumeBucketsAndMaybeBreak(toMerge.size() -1)`. You are "freeing" buckets here, not really making more.